### PR TITLE
feat(webserver): Implement logic for license seat count

### PR DESCRIPTION
**Pull request by boxbeam**

### CODE REVIEW

File: ee/tabby-webserver/src/service/license.rs [hunk 1]
```diff
@@ -59,18 +60,40 @@ fn jwt_timestamp_to_utc(secs: i64) -> Result<DateTime<Utc>> {
...
 struct LicenseServiceImpl {
     db: DbConn,
+    seats: RwLock<(DateTime<Utc>, usize)>,
+}
+
+impl LicenseServiceImpl {
+    async fn seats(&self) -> Result<usize> {
+        let now = Utc::now();
+        let lock = self.seats.read().await;
+        let (refreshed, seats) = &*lock;
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/license.rs:70]: always return `seats`.

But when cache expired, spawn a task to update seat in a background job

File: ee/tabby-webserver/src/service/license.rs [hunk 2]
```diff
@@ -59,18 +60,40 @@ fn jwt_timestamp_to_utc(secs: i64) -> Result<DateTime<Utc>> {
 
 struct LicenseServiceImpl {
     db: DbConn,
+    seats: RwLock<(DateTime<Utc>, usize)>,
+}
+
+impl LicenseServiceImpl {
+    async fn seats(&self) -> Result<usize> {
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/license.rs:67]: read_used_seats

File: ee/tabby-db/src/server_setting.rs [hunk 3]
```diff
@@ -105,4 +105,11 @@ impl DbConn {
         .await?;
         Ok(())
     }
+
+    pub async fn count_seats(&self) -> Result<usize> {
```
Comment (wsxiaoys) [inline ee/tabby-db/src/server_setting.rs:109]: put inside of user.rs, name as count_users

File: ee/tabby-webserver/src/service/license.rs [hunk 4]
```diff
@@ -59,18 +60,40 @@ fn jwt_timestamp_to_utc(secs: i64) -> Result<DateTime<Utc>> {
...
 pub fn new_license_service(db: DbConn) -> impl LicenseService {
-    LicenseServiceImpl { db }
+    LicenseServiceImpl {
+        db,
+        seats: Default::default(),
+    }
 }
 
-fn license_info_from_raw(raw: LicenseJWTPayload) -> Result<LicenseInfo> {
+fn license_info_from_raw(raw: LicenseJWTPayload, seats: usize) -> Result<LicenseInfo> {
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/license.rs:89]: add used seat field to LicenseInfo

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

File: ee/tabby-webserver/src/service/license.rs [hunk 5]
```diff
@@ -59,18 +62,48 @@ fn jwt_timestamp_to_utc(secs: i64) -> Result<DateTime<Utc>> {
...
+                *lock = (now, seats);
+            });
+        }
+        Ok(*seats)
+    }
 }
 
 pub fn new_license_service(db: DbConn) -> impl LicenseService {
-    LicenseServiceImpl { db }
+    LicenseServiceImpl {
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/license.rs:93]: invoke read seats here for a warm start?

**Review by wsxiaoys — COMMENTED** [review]
Otherwise LGTM

File: ee/tabby-db/src/users.rs [hunk 6]
```diff
@@ -206,6 +206,13 @@ impl DbConn {
         .await?;
         Ok(())
     }
+
+    pub async fn count_users(&self) -> Result<usize> {
+        let users = query_scalar!("SELECT COUNT(1) FROM users;")
```
Comment (wsxiaoys) [inline ee/tabby-db/src/users.rs:211]: should only count active users.

File: ee/tabby-db/src/users.rs [hunk 7]
```diff
@@ -206,6 +206,13 @@ impl DbConn {
         .await?;
         Ok(())
     }
+
+    pub async fn count_users(&self) -> Result<usize> {
```
Comment (wsxiaoys) [inline ee/tabby-db/src/users.rs:210]:
```suggestion
    pub async fn count_active_users(&self) -> Result<usize> {
```

File: ee/tabby-db/src/users.rs [hunk 8]
```diff
@@ -206,6 +206,13 @@ impl DbConn {
         .await?;
         Ok(())
     }
+
```
Comment (wsxiaoys) [inline ee/tabby-db/src/users.rs:209]:
```suggestion
// FIXME(boxbeam): consider implement this with a in-memory caching in future.
```

**Review by wsxiaoys — COMMENTED** [review]
Otherwise LGTM

File: ee/tabby-db/src/users.rs (hunk 8 above)
Comment (boxbeam) [reply ee/tabby-db/src/users.rs:209]: But I did implement it with in-memory caching? Not sure what is expected here. The caching just isn't done at the db level.

File: ee/tabby-webserver/src/service/license.rs [hunk 9]
```diff
@@ -62,18 +62,52 @@ fn jwt_timestamp_to_utc(secs: i64) -> Result<DateTime<Utc>> {
 
 struct LicenseServiceImpl {
     db: DbConn,
+    seats: Arc<RwLock<(DateTime<Utc>, usize)>>,
+}
+
+impl LicenseServiceImpl {
+    // TODO: Discuss caching strategies when caching is needed for other fields.
+    // We may want to cache at a different layer, and probably want an abstraction over
+    // this kind of caching logic.
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/license.rs:71]:
```suggestion
```

File: ee/tabby-db/src/users.rs (hunk 8 above)
Comment (wsxiaoys) [inline ee/tabby-db/src/users.rs:209]:
```suggestion
    // FIXME(boxbeam): Revisit if a caching layer should be put into DbConn for this query in future.
```

File: ee/tabby-webserver/src/schema/license.rs [hunk 10]
```diff
@@ -1,10 +1,9 @@
+use anyhow::Result;
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/schema/license.rs:1]: > Use schema::Result as API interface

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

File: ee/tabby-webserver/src/service/license.rs [hunk 11]
```diff
@@ -62,18 +63,49 @@ fn jwt_timestamp_to_utc(secs: i64) -> Result<DateTime<Utc>> {
...
+        let (refreshed, seats) = &*lock;
+        if now - refreshed > Duration::minutes(5) {
+            let seats_clone = self.seats.clone();
+            let db = self.db.clone();
+            tokio::spawn(async move {
+                let mut lock = seats_clone.write().await;
+                let seats = match db.count_active_users().await {
+                    Ok(s) => s,
+                    Err(e) => {
+                        eprintln!("Failed to count users: {e}");
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/license.rs:82]:
```suggestion
                        warn!("Failed to count users: {e}");
```

**Pull request merged**